### ISSUE TITLE
[Feature Request] Per user sync & authentication

### ISSUE THREAD

**Issue by EnergoStalin** [vosmiic/jellyfin-ani-sync#54]
This will be ideal for independent configuration #53 and multi user jellyfin setup.

Also imagine how cool be when you watching anime with syncplay feature and it will update progress for all participants.

**Comment by vosmiic** [comment]
This should be already supported; each user can authenticate using a single app (if its for MyAnimeList or AniList). 
Each access token is stored separately in the users individual config, while the client ID and secret are stored "globally" so that multiple users can authenticate using a single app. This is generally the accepted flow when using app-based API's.
Are you asking if we can have multiple app's (client ID and secret) in a single install? Apologies if I am mis-understanding your query.
>Also imagine how cool be when you watching anime with syncplay feature and it will update progress for all participants.

Agreed, I occasionally use Jellyfin in sync groups and am annoyed that it does not update everyone's status. Unfortunately, I am not able to find a fix to this and I believe it may be an issue on Jellyfin's end (but don't quote me; I have not spent any large amount of time investigating it).
We run through each user when play has completed here:
https://github.com/vosmiic/jellyfin-ani-sync/blob/773c22f8b98fb132d7434fcb1c77de7335741e12/SessionServerEntry.cs#L48-L50
But it seems to only work for the last user to join (at least that's what I found).

**Comment by EnergoStalin** [comment]
Yes you got me right. But in this case would be great to move authenticate button to user profile settings. Because for now it's only possible to authenticate through dashboard which requires all user access to server management. A bit insecure don't you think?

**Comment by vosmiic** [comment]
>But in this case would be great to move authenticate button to user profile settings. Because for now it's only possible to authenticate through dashboard which requires all user access to server management. A bit insecure don't you think?

Agreed, it is insecure and it is annoying to authenticate new users since an admin has to generate the URL and then hand it out to users to authenticate. 
Unfortunately I'm not sure if Jellyfin plugins are able to edit the user settings page, during early development I had a look around and could not find any plugins that did so. 
As an alternative idea, would an API call that returns an authentication URL help?
So something like `/AniSync/userAuthentication?username=fred&provider=AniList` and if a client ID and secret has been supplied for the chosen provider, it would return a URL that the user would click and then become authenticated.
It would skip requiring the user (or you, if you give permissions to your users to generate the URL) having to visit the admin dashboard.
Or we could go down the (arguably simpler and easier to implement) route of environment variables. Either method is fine by me.

**Comment by EnergoStalin** [comment]
> /AniSync/userAuthentication?username=fred&provider=AniList

Implementing such route sounds good, but quoting possibility of using generated auth url, it's maybe unnecessary depending on method of matching users between provider and jellyfin(i'm assuming its done by name from smth like /me request to provider with token like it's done in oauth). If it works like mentioned above then it's already enough while we don't found way inject button to user profile.
>Agreed, it is insecure and it is annoying to authenticate new users since an admin has to generate the URL and then hand it out to users to authenticate.

I think environment variables support would be great anyway especially for docker jellyfin distributions for clean deploy.
> Or we could go down the (arguably simpler and easier to implement) route of environment variables.

**Comment by vosmiic** [comment]
Hey, sorry for not responding here in quite some time. Been working on other projects and assumed we had come to a decision. I presumed the environment variables method would work, but quickly realised we would be back to the original problem since you would be able to setup the authentication details (client ID, secret, callback URL etc.) but you wouldn't be able to set an actual access token (since this requires the Jellyfin web server to fetch, unless you manually get the access token which is a pain). 

>I thinked a little bit more about it. Implementing route like this saving jellyfin - provider user junction and then, match accounts by it in /authCallback would be great idea. this is allow us to have different names for accounts linked and maybe multiple accounts linked to the same providers.

The way that we currently handle the authentication is that when the user clicks on the authorize button and a URL is generated, the username being authenticated on is stored in the plugin details. 
When the user completes authentication and is then redirected back to the blank page (`/authCallback`) the plugin looks up that stored username and binds the access token to its stored configuration details.
This essentially only allows a single user to authenticate at one time. You can't generate multiple user authentication redirect URLs, they must be consumed one after another otherwise the access token will be bound to the last selected user.
As far as I can tell there isn't a nicer way of handling this, since OAuth redirect URLs do not support custom parameters.
So this method could work, but it will only work if each user authenticates in the correct order (`user 1 generates auth URL > user 1 completes authentication > user 2 generates auth URL > user 2 completes authentication`, while `user 1 generates auth URL > user 2 generates auth URL > user 1 completes authentication` would not work since user 1's access token will be bound to users 2's configuration)

So I believe we have 2 potential methods, both with distinct drawbacks: 
1. Environment variables, which don't really solve much as the user/admin will still have to navigate to the plugin settings to generate an authentication URL.
2. Separate authentication URL which will strictly require the users to authenticate one after another to prevent access tokens being bound to incorrect users.

**Comment by EnergoStalin** [comment]
## TLDR:
### Note on previous comment 
It will work if on auth callback you query something like `/me` endpoint and obtain username then looking up for pending auth for user with this username that what I mean previously.

### Name mismatch 
In case of provider and jellyfin usernames not match make possible specify alias for user during queueing step. And maybe considering make aliases more valuable than actual usernames.

### Security 
To avoid account linking stealing store junction in map overwriting keys making a only last auth request valuable(still a bit insecure but we can't do something with that until custom parameters not supported the only way to implicitly provide security is make user input token manually but it's not well suits to this case) we still can have good security by restrict users use provider usernames in jellyfin account but then it will luck flexibility with multiple providers.

## So my suggestions is:
1. Authenticate user by query something like `/me` api endpoint.
2. Restrict users to have matching usernames in jellyfin and provider or implement method described above which a bit insecure though only if we not force users input token manually somewhere or somehow maybe even hosting separate service for it on different port if jellyfin not allow us to define additional pages for that

**Comment by vosmiic** [comment]
>It will work if on auth callback you query something like /me endpoint

That would be quite difficult to implement, as the calls are all unauthenticated. Also the redirect URL comes with no authentication, so there is no way the endpoint knows who it is authenticating.

>Restrict users to have matching usernames in jellyfin and provider or implement method described above which a bit insecure though only if we not force users input token manually somewhere or somehow maybe even hosting separate service for it on different port if jellyfin not allow us to define additional pages for that

Hmm I'm not very keen to add that matching username restriction considering it would cause breaking changes to exsiting users of the plugin. 
Also the separate service is a bit out of scope of this plugin, and would probably be something good to develop separately (probably not by me since I am already working on a lot of different projects), ultimately I think it would be best to make changes to Jellyfin itself and allow plugins to add options to user settings which I am not currently able to do (as stated, already working on a lot of projects).
In theory you should be able to create a very basic application that generates the URL for you since the plugin simply communicates with the backend via controllers, but not something that I am going to look into at the moment.
